### PR TITLE
core: support Mat 1D

### PR DESCRIPTION
**Pull request by alalek**
**WIP**

<cut/>

- [ ] UMat alignment
- [ ] compatibility for `calcHist()` (returns 1D Mat, so minMaxLoc() fails due to `(i, 0)` instead of `(0, i)` results)

```
force_builders=Linux x64,Linux OpenCL,Linux x64 Debug
```

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 1]
```diff
@@ -545,7 +545,7 @@ Mat getMatFromTensor(opencv_onnx::TensorProto& tensor_proto)
         CV_Error(Error::StsUnsupportedFormat, "Unsupported data type: " +
                         opencv_onnx::TensorProto_DataType_Name(datatype));
     if (tensor_proto.dims_size() == 0)
-        blob.dims = 1;  // To force 1-dimensional cv::Mat for scalars.
+        CV_CheckEQ(blob.dims, 1, "Tensor is scalar");  // ensure 1-dimensional cv::Mat for scalars
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:548]: Current list of failed DNN tests require better and accurate handing of shapes in ONNX importer and probably in layers implementations.

We should use 0D tensor (scalar) here. 1D doesn't work properly.

**Comment by vpisarev** [comment]
@fengyuentau, in the following comments I will mark the parts can be moved into a separate PR and polished/integrated separately.

File: modules/core/src/arithm.cpp [hunk 2]
```diff
@@ -188,7 +188,7 @@ static void binary_op( InputArray _src1, InputArray _src2, OutputArray _dst,
 
     if( dims1 <= 2 && dims2 <= 2 && kind1 == kind2 && sz1 == sz2 && type1 == type2 && !haveMask )
     {
-        _dst.create(sz1, type1);
+        _dst.createSameSize(*psrc1, type1);
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp:191]: separate PR

File: modules/core/src/arithm.cpp [hunk 3]
```diff
@@ -1242,7 +1242,7 @@ void cv::compare(InputArray _src1, InputArray _src2, OutputArray _dst, int op)
     if( kind1 == kind2 && src1.dims <= 2 && src2.dims <= 2 && src1.size() == src2.size() && src1.type() == src2.type() )
     {
         int cn = src1.channels();
-        _dst.create(src1.size(), CV_8UC(cn));
+        _dst.createSameSize(src1, CV_8UC(cn));
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp:1245]: separate PR

File: modules/core/src/convert.dispatch.cpp [hunk 4]
```diff
@@ -188,18 +188,33 @@ void Mat::convertTo(OutputArray _dst, int _type, double alpha, double beta) cons
...
         copyTo(_dst);
         return;
     }
 
-    Mat src = *this;
-    if( dims <= 2 )
-        _dst.create( size(), _type );
-    else
-        _dst.create( dims, size, _type );
+    _dst.create(src.dims, src.size, _type);
```
Comment (vpisarev) [inline modules/core/src/convert.dispatch.cpp:217]: separate PR

File: modules/core/src/minmax.cpp [hunk 5]
```diff
@@ -1562,11 +1562,22 @@ void cv::minMaxLoc( InputArray _img, double* minVal, double* maxVal,
 {
     CV_INSTRUMENT_REGION();
 
-    CV_Assert(_img.dims() <= 2);
+    int dims = _img.dims();
```
Comment (vpisarev) [inline modules/core/src/minmax.cpp:1565]: separate PR

File: modules/core/src/minmax.cpp [hunk 6]
```diff
@@ -1562,11 +1562,22 @@ void cv::minMaxLoc( InputArray _img, double* minVal, double* maxVal,
...
+    CV_Assert(dims <= 2);
 
     minMaxIdx(_img, minVal, maxVal, (int*)minLoc, (int*)maxLoc, mask);
-    if( minLoc )
-        std::swap(minLoc->x, minLoc->y);
-    if( maxLoc )
-        std::swap(maxLoc->x, maxLoc->y);
+    if (minLoc)
```
Comment (vpisarev) [inline modules/core/src/minmax.cpp:1569]: separate PR

File: modules/core/test/test_mat.cpp [hunk 7]
```diff
@@ -2356,4 +2356,95 @@ TEST(Mat, regression_18473)
 }
 
 
+TEST(Mat1D, basic)
```
Comment (vpisarev) [inline modules/core/test/test_mat.cpp:2359]: the test would be useful to move to separate PR, but mark it as disabled for now

**Comment by fengyuentau** [comment]
Thank you @vpisarev ! Let me take a look into the code.
